### ISSUE TITLE
`open` of directory without `O_DIRECTORY` does not allow `getdents`

### ISSUE THREAD

**Issue by batrick** [cooperative-computing-lab/cctools#1104]
If a directory is not opened with `O_DIRECTORY`, then Parrot does not allow getting its entries via `getdents`. It will return `ENOTDIR`.

`pfs_table.cc` only opens directories with this flag:

https://github.com/cooperative-computing-lab/cctools/blob/master/parrot/src/pfs_table.cc#L597-L615

`O_DIRECTORY` was not meant to be interpreted this way. The man page says:

````
       O_DIRECTORY
              If pathname is not a directory, cause the open to fail.  This flag was added in kernel version 2.1.126, to avoid  denial-of-service  problems
              if opendir(3) is called on a FIFO or tape device.
````

So it's only intended to make `open` fail if the file is not a directory.

Right now I think the fix may be to try `pname.service->getdir(&pname)` first. If that fails with `ENOTDIR`, then proceed with a regular `pname.service->open(&pname,flags,mode)`. What do you think @dthain?

This was reported by Rodney Walker. You can reproduce it using:

`mkdir foo && ./parrot_run -d all -o debug /bin/tar -czf foo.tgz foo/`

**Comment by batrick** [comment]
@dthain, this fixes Rodney's problem:

https://github.com/batrick/cctools/commit/415315ef65610f2a1d70d91cb19fa34a6296e768

but I think the O_DIRECTORY logic in pfs_table.cc is still wrong. If tar were executed on a directory in Chirp, we get this:

````
$ rm foo.tgz debug; ./parrot_run -d all -o debug /bin/tar czf foo.tgz /chirp/disc01.crc.nd.edu/
/bin/tar: Removing leading `/' from member names
/bin/tar: /chirp/disc01.crc.nd.edu: Cannot open: Is a directory
/bin/tar: Exiting with failure status due to previous errors
$ grep -C5 'Is a directory' debug 
2016/01/18 10:39:45.41 parrot_run[3398] <child:3400> syscall: mprotect
2016/01/18 10:39:45.41 parrot_run[3398] <child:3399> syscall: openat
2016/01/18 10:39:45.41 parrot_run[3398] <child:3399> debug: complete_at_path: `/chirp/disc01.crc.nd.edu' -> `/chirp/disc01.crc.nd.edu'
2016/01/18 10:39:45.41 parrot_run[3398] <child:3399> libcall: open /chirp/disc01.crc.nd.edu 657664 0
2016/01/18 10:39:45.41 parrot_run[3398] <child:3399> resolve: /chirp/disc01.crc.nd.edu = /chirp/disc01.crc.nd.edu,5
2016/01/18 10:39:45.41 parrot_run[3398] <child:3399> libcall: = -1 Is a directory [pfs_open]
2016/01/18 10:39:45.41 parrot_run[3398] <child:3400> syscall: = 0 [mprotect]
2016/01/18 10:39:45.41 parrot_run[3398] <child:3399> syscall: = -21 Is a directory [openat]
````

**Comment by dthain** [comment]
I recall this issue from some years back, but my memory is fuzzy, so I suggest caution...

Suggest that you look at how Linux deals with the all the combinations of open/opendir with a file/directory to make sure that we are accepting/rejecting with the right errno in each case.

**Comment by batrick** [comment]
@trshaffer, as we discussed, please fork this branch:

https://github.com/batrick/cctools/tree/pfs-dir-bugfix

Expand on the test case if necessary and correct the `O_DIRECTORY` logic so that it works for all services (my current branch only fixes the local service). Make sure that cvmfs/irods/chirp do not break because of any changes.

**Comment by batrick** [comment]
Fixed by #1126.
